mgmt: hawkbit: define attributes in the user application

maass-hamburg opened the pull request:
Attributes can now be defined in the user application by the use of a callback function. 
Resolves: #37640 

![image](https://github.com/zephyrproject-rtos/zephyr/assets/56227405/78eaf838-dc5e-4700-a62c-ccfc5fe5f3f8)

maass-hamburg commented:
ping

maass-hamburg commented:
ping @ycsin

maass-hamburg commented:
just a rebase to the current main

ycsin commented:
I haven't really looked at this in details yet but it seems to expose a little too much to the public than I would like, i.e. all the structs and stuff that used to be internal.

I wonder if it is possible to just have a mechanism for hawkBit to get the formatted json of the attributes from the app in the form of a buffer and sent it to the server?

maass-hamburg commented:
> I haven't really looked at this in details yet but it seems to expose a little too much to the public than I would like, i.e. all the structs and stuff that used to be internal.

they are needed if the user wants to define its own config callback. but maybe they could be put in a seperate header file and not in `include/zephyr/mgmt/hawkbit.h` maybe `include/zephyr/mgmt/hawkbit_config.h` or create a new folder `include/zephyr/mgmt/hawkbit/` which contains `include/zephyr/mgmt/hawkbit/hawkbit.h` and `include/zephyr/mgmt/hawkbit/config.h` ?
 
> I wonder if it is possible to just have a mechanism for hawkBit to get the formatted json of the attributes from the app in the form of a buffer and sent it to the server?

It might be, but then we can't use the json library to encode `struct json_obj_descr json_cfg_descr`, only for `struct json_obj_descr json_cfg_data_descr`. As it is currently not possible to add a already encoded object to a `json_obj_descr`.

maass-hamburg committed:
mgmt: hawkbit: add support to define attributes

Allows the attributes to be defined in the user application
by using a callback function.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
samples: mgmt: hawkbit: add custom attributes

Extend sample for the custom attributes

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg commented:
@d3zd3z should I put the change of the json lib into a seperate pr?

ycsin commented on subsys/mgmt/hawkbit/hawkbit.c:
```diff
@@ -827,6 +857,7 @@ static bool send_request(enum http_method method, enum hawkbit_http_request type
 	hb_context.http_req.header_fields = (const char **)headers;
 #endif
 	hb_context.final_data_received = false;
+	hb_context.status_buffer_size = sizeof(hb_context.status_buffer);
```
Good catch! This looks like a potential bug fix and should probably be in another PR, so that we can backport it to earlier releases.

I also wonder if it make sense to remove `status_buffer_size` from the struct and simply use `sizeof(hb_context.status_buffer) - 1` instead of `hb_context.status_buffer_size - 1`?

maass-hamburg replied:
this is now addressed in #70355

ycsin reviewed (commented):
> @d3zd3z should I put the change of the json lib into a seperate pr?

It might be a good idea to create a PR for that change and add proper tests/samples for it so that it is easier to review

ycsin commented on subsys/mgmt/hawkbit/hawkbit.c:
```diff
@@ -840,10 +871,18 @@ static bool send_request(enum http_method method, enum hawkbit_http_request type
 
 	case HAWKBIT_CONFIG_DEVICE:
-		memset(&cfg, 0, sizeof(cfg));
+		struct hawkbit_cfg cfg = {0};
```
What do you think about sending only the minimal info in default hawkBit implementation (i.e. the existing stuff like `VIN` (I forgot if `hwRevision` is strictly required)), and `merge` the custom attributes in an additional steps?

maass-hamburg replied:
`VIN` and `hwRevision` are both not strictly required by hawkBit. But I would still like to send `VIN` as our default configuration. `hwRevision` will be removed from the default configuration.

ycsin commented on subsys/mgmt/hawkbit/hawkbit.c:
```diff
@@ -840,10 +871,18 @@ static bool send_request(enum http_method method, enum hawkbit_http_request type
+
+		ret = hawkbit_config_device_data_cb_handler(device_id, cfg_data, sizeof(cfg_data));
+		if (ret) {
+			LOG_ERR("Can't encode the JSON script (HAWKBIT_CONFIG_DEVICE, data): %d",
```
Please try to adopt the pattern of #69229 if possible

ycsin reviewed (commented):
probably needs a rebase

m-byte commented on subsys/mgmt/hawkbit/hawkbit.c:
```diff
@@ -44,7 +44,7 @@ LOG_MODULE_REGISTER(hawkbit, CONFIG_HAWKBIT_LOG_LEVEL);
 #define URL_BUFFER_SIZE 300
 #define SHA256_HASH_SIZE 32
-#define STATUS_BUFFER_SIZE 200
+#define STATUS_BUFFER_SIZE (100 + CONFIG_HAWKBIT_ADDITIONAL_STATUS_BUFFER_SIZE)
```
I feel like it would be better to move the whole STATUS_BUFFER_SIZE to Kconfig. Otherwise, there is this weird semi-magic number.
Maybe add a recommendation to the help text so users know that they shouldn't set that value below 100

maass-hamburg replied:
the 100 was just the half of the default value, it can also be set below 100

maass-hamburg commented:
@ycsin With #70383 I want to go back to let the user encode `struct json_obj_descr json_cfg_descr` instead of just `struct json_obj_descr json_cfg_data_descr`, because it only contains of `data` and `mode` and no internal stuff has to be exposed. So the user can also set the mode and we don't need the change in the json lib.

maass-hamburg commented:
just a rebase to resolve conflicts

ycsin commented on subsys/mgmt/hawkbit/Kconfig:
```diff
@@ -79,6 +79,19 @@ config HAWKBIT_DDI_SECURITY_TOKEN
+	  Use custom definition of device attributes.
+
+config HAWKBIT_STATUS_BUFFER_SIZE
+	int "Hawkbit status buffer size"
```
```suggestion
	int "hawkBit status buffer size"
```

ycsin reviewed (commented):
The approach in this PR should work, I just wonder if we can keep the default implementation simple (only send `device_id`) but provide more flexibility by having a mechanism (another step or public function, if enabled) to "merge" any custom attributes to the record in the server, i.e. the reason of last reboot (reset cause), die temp, stack watermark, ...

ycsin commented on samples/subsys/mgmt/hawkbit/prj.conf:
```diff
@@ -58,3 +58,6 @@ CONFIG_LOG_BUFFER_SIZE=4096
 #Generate HEX output
 CONFIG_BUILD_OUTPUT_HEX=y
+
+#Use custom attributes for Hawkbit
```
hawkBit

maass-hamburg commented:
> The approach in this PR should work, I just wonder if we can keep the default implementation simple (only send `device_id`) but provide more flexibility by having a mechanism (another step or public function, if enabled) to "merge" any custom attributes to the record in the server.

As sending the `device_id` as `VIN` is not required by hawkBit. hawkBit doesn't use for anything special. If the user doesn't want to send that, he should be able to do so.

maass-hamburg commented:
> The approach in this PR should work, I just wonder if we can keep the default implementation simple (only send `device_id`) but provide more flexibility by having a mechanism (another step or public function, if enabled) to "merge" any custom attributes to the record in the server, i.e. the reason of last reboot (reset cause), die temp, stack watermark, ...

This problem is that the device attributes are only send to the server if the server requests it. This normally only happens once. the hawkBit server will only request it again, if you go in the user interface and click there on `Request attributes update`. It could be done to send it without the server requesting but that's outside of the scope of this PR .

ycsin commented on include/zephyr/mgmt/hawkbit.h:
```diff
@@ -36,6 +36,30 @@ enum hawkbit_response {
+typedef int (*hawkbit_config_device_data_cb_handler_t)(const char *device_id, uint8_t *buffer,
+						  const size_t buffer_size);
+
+/**
+ * @brief Set the custom data callback.
+ * @details This function is used to set the custom data callback.
+ * The callback is used to provide the custom data to the hawkBit server.
+ * @param cb The callback function.
+ * @retval 0 on success.
+ * @retval -EINVAL if the callback is NULL.
+ */
```
We usually have an empty line between `brief`, `details`, `param` and `return`/`retval`, but since `return` is already used in `hawkbit_probe`, it should be used here as well for consistency

maass-hamburg replied:
is now changed

ycsin reviewed (commented):
This patch works - Have you considered doing this kind of one time configurations during `hawkbit_init` by extending it to accept an config struct? The same idea can also applies to #68806

maass-hamburg commented:
> This patch works - Have you considered doing this kind of one time configurations during `hawkbit_init` by extending it to accept an config struct? The same idea can also applies to #68806

I would prefer to not do it that way, as most subsystems in zephyr are also using this separate approach when it comes to setting callbacks.

ycsin reviewed (dismissed):
The doc can use some [improvements](https://github.com/zephyrproject-rtos/zephyr/pull/69293/files#r1561963951), LGTM in general

ycsin commented on samples/subsys/mgmt/hawkbit/src/main.c:
```diff
@@ -20,6 +21,40 @@
+	struct hawkbit_cfg cfg = {
+		.mode = "merge",
+		.data.VIN = device_id,
+		.data.customAttr = "Hello World!",
```
good to attach a screenshot of this in the web UI

maass-hamburg replied:
![image](https://github.com/zephyrproject-rtos/zephyr/assets/56227405/78eaf838-dc5e-4700-a62c-ccfc5fe5f3f8)

maass-hamburg commented:
rebased to resolve conflicts

ycsin reviewed (approved):
Re-ack

carlescufi reviewed (approved):
LGTM
